homeassistant/components/config/device_registry.py

```diff
@@ -70,6 +68,52 @@ async def websocket_update_device(hass, connection, msg):
     connection.send_message(websocket_api.result_message(msg_id, _entry_dict(entry)))
 
 
+@websocket_api.require_admin
+@websocket_api.websocket_command(
+    {
+        "type": "config/device_registry/remove_config_entry",
+        "device_id": str,
+        "config_entry_id": str,
+    }
+)
+@callback
+def websocket_remove_config_entry_from_device(
+    hass: HomeAssistant, connection: websocket_api.ActiveConnection, msg: dict
+) -> None:
+    """Remove config entry from a device."""
+    registry = async_get(hass)
+
+    config_entry = hass.config_entries.async_get_entry(msg["config_entry_id"])
+    if config_entry and not config_entry.supports_remove_device:
+        connection.send_error(
+            msg["id"],
+            websocket_api.const.ERR_HOME_ASSISTANT_ERROR,
+            "Config entry does not support device removal",
+        )
+        return
+
+    try:
+        entry = registry.async_update_device(
```

Comment:
Didn't we still want to call a callback in the integration before removing the config entry from the device? We didn't decide to remove that feature, as I interpreted the conversation, just move everything here to the device registry.

Reply:
The problem is that if we do that, we can't migrate the logic to `DeviceRegistry.async_update_device` because it requires calls to coroutine functions to get hold of the component.
Hence I suggest that integrations subscribe to `EVENT_DEVICE_REGISTRY_UPDATED` instead and act if they've been removed from a device.

Reply:
Why should we do it there? We have a websocket command handler.

Reply:
The reason I think an integration callback should run first is how I view the responsibilities. The device registry should represent the state of the device in the integration. Ie, the truth comes from the integration.

If the user wants to remove a device, the integration should be asked to remove the device and then we update the device registry based on the result in the integration.

Reply:
You can use `getattr(hass.components, domain)` and it's usable in a callback 😬

Reply:
> use getattr(hass.components, domain) and it's usable in a callback

Is that OK, there's no need to do like this?
```py
        try:
            integration = await loader.async_get_integration(
                self.hass, config_entry.domain
            )
        except loader.IntegrationNotFound as exc:
            raise HomeAssistantError("Integration not found") from exc
        try:
            component = integration.get_component()
        except ImportError as exc:
            raise HomeAssistantError("Component not found") from exc
```

Reply:
> Why should we do it there? We have a websocket command handler.

If we trust the calls to `async_update_device` to have done all necessary checks already it's fine to just check in the websocket command handler (which we for sure can't trust).

Reply:
> we update the device registry based on the result in the integration

So we not only want to let the integration know, we are back to allowing it to veto being removed from the device?
@balloob you also agree this is the way to go, and this PR can be finished?

Reply:
yes